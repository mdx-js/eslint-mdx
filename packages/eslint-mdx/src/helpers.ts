/// <reference path="../typings.d.ts" />

import type { Linter } from 'eslint'
import type { Position as ESPosition, SourceLocation } from 'estree'
import type { Point, Position } from 'unist'

import type {
  Arrayable,
  JsxNode,
  ParserFn,
  ParserOptions,
  ValueOf,
} from './types'

export const FALLBACK_PARSERS = [
  '@typescript-eslint/parser',
  '@babel/eslint-parser',
  'babel-eslint',
  'espree',
] as const

export const JSX_TYPES = ['JSXElement', 'JSXFragment']

export const isJsxNode = (node: { type: string }): node is JsxNode =>
  JSX_TYPES.includes(node.type)

// eslint-disable-next-line sonarjs/cognitive-complexity
export const normalizeParser = (parser?: ParserOptions['parser']) => {
  if (parser) {
    if (typeof parser === 'string') {
      // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
      parser = require(parser) as ParserOptions['parser']
    }

    if (typeof parser === 'object') {
      parser =
        ('parseForESLint' in parser && parser.parseForESLint) ||
        ('parse' in parser && parser.parse)
    }

    if (typeof parser !== 'function') {
      throw new TypeError(`Invalid custom parser for \`eslint-mdx\`: ${parser}`)
    }

    return [parser]
  }

  const parsers: ParserFn[] = []

  // try to load FALLBACK_PARSERS automatically
  for (const fallback of FALLBACK_PARSERS) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
      const fallbackParser = require(fallback) as Linter.ParserModule
      /* istanbul ignore next */
      const parserFn =
        'parseForESLint' in fallbackParser
          ? // eslint-disable-next-line @typescript-eslint/unbound-method
            fallbackParser.parseForESLint
          : // eslint-disable-next-line @typescript-eslint/unbound-method
            fallbackParser.parse
      /* istanbul ignore else */
      if (parserFn) {
        parsers.push(parserFn)
      }
    } catch {}
  }

  return parsers
}

export interface BaseNode {
  type: string
  loc: SourceLocation
  range: [number, number]
  start?: number
  end?: number
}

export const normalizePosition = (loc: Position): Omit<BaseNode, 'type'> => {
  const start = loc.start.offset
  const end = loc.end.offset
  return {
    range: [start, end],
    loc,
    start,
    end,
  }
}

export const hasProperties = <T, P extends keyof T = keyof T>(
  obj: unknown,
  properties: Arrayable<P>,
): obj is T =>
  typeof obj === 'object' &&
  obj &&
  properties.every(property => property in obj)

// fix #292
export const getPositionAt = (code: string, offset: number): ESPosition => {
  let currOffset = 0

  const lines = code.split('\n')

  // eslint-disable-next-line unicorn/no-for-loop
  for (let index = 0; index < lines.length; index++) {
    const line = index + 1
    const nextOffset = currOffset + lines[index].length

    if (nextOffset >= offset) {
      return {
        line,
        column: offset - currOffset,
      }
    }

    currOffset = nextOffset + 1 // add a line break `\n` offset
  }
}

export const restoreNodeLocation = <T>(node: T, point: Point): T => {
  if (node && typeof node === 'object') {
    for (const value of Object.values(node) as Array<ValueOf<T>>) {
      restoreNodeLocation(value, point)
    }
  }

  if (!hasProperties<BaseNode>(node, ['loc', 'range'])) {
    return node
  }

  let {
    loc: { start: startLoc, end: endLoc },
    range: [start, end],
  } = node

  const range = [(start += point.offset), (end += point.offset)] as const

  return Object.assign(node, {
    start,
    end,
    range,
    loc: {
      start: {
        line: point.line + startLoc.line,
        column: startLoc.column + (startLoc.line === 1 ? point.column : 0),
      },
      end: {
        line: point.line + endLoc.line,
        column: endLoc.column + (endLoc.line === 1 ? point.column : 0),
      },
    },
  })
}

export const arrayify = <T, R = T extends Array<infer S> ? S : T>(
  ...args: T[]
) =>
  args.reduce<R[]>((arr, curr) => {
    arr.push(...(Array.isArray(curr) ? curr : curr == null ? [] : [curr]))
    return arr
  }, [])

export const first = <T>(items: T[] | readonly T[]) => items && items[0]

export const last = <T>(items: T[] | readonly T[]) =>
  items && items[items.length - 1]
