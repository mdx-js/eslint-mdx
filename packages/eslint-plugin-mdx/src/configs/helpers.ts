export const getGlobals = <
  T extends Record<string, unknown> | string[],
  G extends Record<string, boolean>,
  R extends G = G & Record<T extends Array<infer R> ? R : keyof T, false>
>(
  sources: T,
  initialGlobals: G = {} as G,
): R =>
  (Array.isArray(sources)
    ? (sources as string[])
    : Object.keys(sources)
  ).reduce(
    (globals, source) =>
      Object.assign(globals, {
        [source]: false,
      }),
    initialGlobals as R,
  )
