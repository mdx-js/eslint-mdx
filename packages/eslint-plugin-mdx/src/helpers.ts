export const getGlobals = <
  T extends Record<string, unknown> | string[],
  G extends Record<string, boolean>,
  R extends G = G & Record<T extends Array<infer R> ? R : keyof T, false>,
>(
  sources: T,
  initialGlobals: G = {} as G,
): R =>
  (Array.isArray(sources)
    ? // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
      (sources as string[])
    : Object.keys(sources)
  ).reduce(
    (globals, source) =>
      Object.assign(globals, {
        [source]: false,
      }),
    // eslint-disable-next-line @typescript-eslint/prefer-reduce-type-parameter
    initialGlobals as R,
  )
