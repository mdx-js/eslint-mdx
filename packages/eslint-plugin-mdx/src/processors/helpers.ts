import { last } from 'eslint-mdx'

export const DEFAULT_LANGUAGE_MAPPER: Record<string, string> = {
  javascript: 'js',
  javascriptreact: 'jsx',
  typescript: 'ts',
  typescriptreact: 'tsx',
  markdown: 'md',
  mdown: 'md',
  mkdn: 'md',
}

export function getShortLang(
  filename: string,
  languageMapper?: Record<string, string> | false,
): string {
  const language = last(filename.split('.'))
  if (languageMapper === false) {
    return language
  }
  languageMapper = { ...DEFAULT_LANGUAGE_MAPPER, ...languageMapper }
  const lang = language.toLowerCase()
  return languageMapper[language] || languageMapper[lang] || lang
}
