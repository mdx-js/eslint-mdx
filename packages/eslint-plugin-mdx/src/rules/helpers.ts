import path from 'path'

import { cosmiconfigSync } from 'cosmiconfig'
import type { CosmiconfigResult } from 'cosmiconfig/dist/types'
import remarkMdx from 'remark-mdx'
import remarkParse from 'remark-parse'
import remarkStringify from 'remark-stringify'
import unified from 'unified'

import type { RemarkConfig } from './types'

export const requirePkg = <T>(
  plugin: string,
  prefix: string,
  filePath?: string,
): T => {
  if (filePath && /^\.\.?([/\\]|$)/.test(plugin)) {
    plugin = path.resolve(path.dirname(filePath), plugin)
  }
  prefix = prefix.endsWith('-') ? prefix : prefix + '-'
  const packages = [
    plugin,
    plugin.startsWith('@')
      ? plugin.replace('/', '/' + prefix)
      : prefix + plugin,
  ]
  let error: Error
  for (const pkg of packages) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
      return require(pkg) as T
    } catch (err) {
      if (!error) {
        error = err as Error
      }
    }
  }
  throw error
}

let searchSync: (searchFrom?: string) => CosmiconfigResult

export const remarkProcessor = unified().use(remarkParse).freeze()

export const getRemarkProcessor = (searchFrom: string, isMdx: boolean) => {
  if (!searchSync) {
    searchSync = cosmiconfigSync('remark', {
      packageProp: 'remarkConfig',
    }).search
  }

  /* istanbul ignore next */
  let result: Partial<CosmiconfigResult> = {}

  try {
    result = searchSync(searchFrom)
  } catch (err) {
    // https://github.com/eslint/eslint/issues/11989
    /* istanbul ignore if */
    if (
      (err as { code?: string }).code !== 'ENOTDIR' ||
      !/\d+_?\.[a-z]+$/.test(searchFrom)
    ) {
      throw err
    }
    try {
      result = searchSync(path.dirname(searchFrom))
    } catch {
      /* istanbul ignore next */
      throw err
    }
  }

  /* istanbul ignore next */
  const { plugins = [], settings } = (result.config ||
    {}) as Partial<RemarkConfig>

  try {
    // disable this rule automatically since we already have a parser option `extensions`
    // eslint-disable-next-line node/no-extraneous-require
    plugins.push([require.resolve('remark-lint-file-extension'), false])
  } catch {
    // just ignore if the package does not exist
  }

  const initProcessor = remarkProcessor().use({ settings }).use(remarkStringify)

  if (isMdx) {
    initProcessor.use(remarkMdx)
  }

  return plugins
    .reduce((processor, pluginWithSettings) => {
      const [plugin, ...pluginSettings] = Array.isArray(pluginWithSettings)
        ? pluginWithSettings
        : [pluginWithSettings]
      return processor.use(
        /* istanbul ignore next */
        typeof plugin === 'string'
          ? requirePkg(plugin, 'remark', result.filepath)
          : plugin,
        ...pluginSettings,
      )
    }, initProcessor)
    .freeze()
}
