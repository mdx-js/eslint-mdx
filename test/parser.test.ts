/* eslint-disable sonarjs/no-duplicate-string */
import type { Node, Parent, ParserConfig, ParserOptions } from 'eslint-mdx'
import {
  DEFAULT_PARSER_OPTIONS as parserOptions,
  first,
  mdxProcessor,
  normalizeParser,
  parser,
} from 'eslint-mdx'

import { noop } from './helpers'

const stringToNode = (text: string) =>
  first((mdxProcessor.parse(text) as Parent).children)

describe('parser', () => {
  it('should transform html style comment in jsx into jsx comment', () => {
    const sourceText = `<Comment><!-- JSX Comment --><!-- JSX Comment --></Comment>`
    const expectText = `<Comment>{/** JSX Comment */}{/** JSX Comment */}</Comment>`
    expect(parser.normalizeJsxNode(stringToNode(sourceText))).toEqual({
      type: 'jsx',
      data: {
        jsxType: 'JSXElementWithHTMLComments',
        comments: [
          {
            fixed: '{/** JSX Comment */}',
            loc: {
              end: {
                column: 29,
                line: 1,
                offset: 29,
              },
              start: {
                column: 9,
                line: 1,
                offset: 9,
              },
            },
            origin: '<!-- JSX Comment -->',
          },
          {
            fixed: '{/** JSX Comment */}',
            loc: {
              end: {
                column: 49,
                line: 1,
                offset: 49,
              },
              start: {
                column: 29,
                line: 1,
                offset: 29,
              },
            },
            origin: '<!-- JSX Comment -->',
          },
        ],
        inline: false,
      },
      value: expectText,
      position: {
        end: {
          column: 60,
          line: 1,
          offset: 59,
        },
        indent: [],
        start: {
          column: 1,
          line: 1,
          offset: 0,
        },
      },
    })
  })

  it('should parse adjacent JSX nodes correctly', () => {
    const sourceText = `<header>Header</header><main>Main Content</main>`

    const parsedNodes = parser.normalizeJsxNode(stringToNode(sourceText))

    expect((parsedNodes as Node[]).length).toBe(2)
    expect(parsedNodes).toMatchObject([
      {
        type: 'jsx',
        value: `<header>Header</header>`,
        position: {
          start: { line: 1, column: 0, offset: 0 },
          end: { line: 1, column: 23, offset: 23 },
        },
      },
      {
        type: 'jsx',
        value: `<main>Main Content</main>`,
        position: {
          start: { line: 1, column: 23, offset: 23 },
          end: { line: 1, column: 48, offset: 48 },
        },
      },
    ])
  })

  it('should throw on invalid parser', () => {
    const parserConfigs: ParserConfig[] = [
      {
        parse: null,
      },
      {
        // @ts-expect-error
        parseForEsLint: null,
      },
    ]
    for (const p of parserConfigs)
      expect(() =>
        parser.parse('<header>Header</header>', {
          ...parserOptions,
          parser: p,
        }),
      ).toThrowErrorMatchingSnapshot()

    expect(() =>
      parser.parse('<header>Header</header>', {
        ...parserOptions,
        parser: noop as ParserOptions['parser'],
      }),
    ).toThrow("Cannot use 'in' operator to search for 'ast' in undefined")
  })

  it('should work with valid custom parser', () => {
    expect(() =>
      parser.parse('<header>Header</header>', {
        ...parserOptions,
        sourceType: undefined,
        parser: '@babel/eslint-parser',
        babelOptions: {
          configFile: require.resolve('@1stg/babel-preset/config'),
        },
      }),
    ).not.toThrow()
  })

  it('should fallback to espree if no preferred parsers found', () => {
    jest
      .setMock('@typescript-eslint/parser', null)
      .setMock('@babel/eslint-parser', null)
      .mock('babel-eslint', noop, { virtual: true })
      .setMock('espree', { parse: noop })
    expect(normalizeParser()).toEqual([noop])
    jest.unmock('@typescript-eslint/parser').unmock('@babel/eslint-parser')
  })

  it('should throw on invalid es syntaxes', () => {
    expect(() =>
      parser.parse("import A from 'a'\nimport A from 'a'", parserOptions),
    ).toThrow("unknown: Identifier 'A' has already been declared")
    expect(() => parser.parse('<header><>\n</header>', parserOptions)).toThrow(
      'Expected corresponding JSX closing tag for <>',
    )
    expect(() => parser.parse('<h1></h2>', parserOptions)).toThrow(
      'Expected corresponding JSX closing tag for <h1>',
    )
    expect(() =>
      parser.parse('Header\n<>', {
        ...parserOptions,
        parser: '@typescript-eslint/parser',
      }),
    ).toThrowError()
    expect(() =>
      parser.parse('Header\n<>', {
        ...parserOptions,
        parser: '@babel/eslint-parser',
      }),
    ).toThrowError()
    expect(() => parser.parse('<main><</main>', parserOptions)).toThrow(
      'Unexpected token (1:10)',
    )
    expect(() => parser.parse('<main>{<}</main>', parserOptions)).toThrow(
      'Unexpected token (1:11)',
    )
    expect(() =>
      parser.parse('<main>\n<section><</section></main>', parserOptions),
    ).toThrow('Unexpected token (2:10)')
  })

  it('should not throw on adjacent JSX nodes', () =>
    expect(() =>
      parser.parse(
        '<header></header>\n<main><section>left</section><section>right<input name="name"/></section></main>',
        parserOptions,
      ),
    ).not.toThrow())

  it('should not throw on JSX with blank lines', () =>
    expect(() =>
      parser.parse('<header>\n\nTitle\n\n</header>', parserOptions),
    ).not.toThrow())

  it("should not throw on <$> or </$> because it's not considered as jsx", () => {
    expect(() => parser.parse('<$>', parserOptions)).not.toThrow()
    expect(() => parser.parse('</$>', parserOptions)).not.toThrow()
  })

  it('should be able to parse normal js file', () => {
    expect(() =>
      parser.parse("import A from 'a'", {
        ...parserOptions,
        filePath: 'test.js',
      }),
    ).not.toThrow()
    expect(() =>
      parser.parse('const a = {}', {
        ...parserOptions,
        filePath: 'test.js',
      }),
    ).not.toThrow()
  })

  it('should work with plain markdown file', () =>
    expect(() =>
      parser.parse('<img><br>', {
        ...parserOptions,
        filePath: 'test.md',
      }),
    ).not.toThrow())

  it('should work with multiple comments', () =>
    expect(() =>
      parser.parse('<!-- * foo -->\n<!-- * bar -->', {
        ...parserOptions,
        filePath: 'test.mdx',
      }),
    ).not.toThrow())
})
