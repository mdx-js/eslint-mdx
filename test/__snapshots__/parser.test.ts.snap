// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should throw on invalid parser 1`] = `"Invalid custom parser for \`eslint-mdx\`: null"`;

exports[`parser should throw on invalid parser 2`] = `"Invalid custom parser for \`eslint-mdx\`: false"`;
